Microsoft.Toolkit/Diagnostics/Generated/Guard.Collection.tt

```diff
@@ -0,0 +1,259 @@
+        public static void HasSizeEqualTo<T>(<#=item.Type#> source, <#=item.DestinationType#> destination, string name)
+        {
+            if (source.<#=item.Size#> != destination.<#=item.Size#>)
+            {
+<#
+    if (item.HasCountProperty)
+    {
+#>
+                ThrowHelper.ThrowArgumentExceptionForHasSizeEqualTo(<#=item.Cast#>source, destination.<#=item.Size#>, name);
+<#
+    }
+    else
+    {
+#>
+                ThrowHelper.ThrowArgumentExceptionForHasSizeEqualTo(source, <#=item.Cast#>destination, name);
+<#
+    }
+#>
+            }
+        }
+
+        /// <summary>
+        /// Asserts that the source <see cref="ReadOnlyMemory{T}"/> instance must have a size of less than or equal to that of a destination <#=item.XmlType#> instance.
+        /// </summary>
+        /// <typeparam name="T">The item of items in the input <#=item.XmlType#> instance.</typeparam>
+        /// <param name="source">The source <#=item.XmlType#> instance to check the size for.</param>
+        /// <param name="destination">The destination <#=item.XmlType#> instance to check the size for.</param>
+        /// <param name="name">The name of the input parameter being tested.</param>
+        /// <exception cref="ArgumentException">Thrown if the size of <paramref name="source"/> is > the one of <paramref name="destination"/>.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void HasSizeLessThanOrEqualTo<T>(<#=item.Type#> source, <#=item.DestinationType#> destination, string name)
+        {
+            if (source.<#=item.Size#> > destination.<#=item.Size#>)
+            {
+<#
+    if (item.HasCountProperty)
+    {
+#>
+                ThrowHelper.ThrowArgumentExceptionForHasSizeEqualTo(<#=item.Cast#>source, destination.<#=item.Size#>, name);
+<#
+    }
+    else
+    {
+#>
+                ThrowHelper.ThrowArgumentExceptionForHasSizeLessThanOrEqualTo(source, <#=item.Cast#>destination, name);
+<#
+    }
+#>
+            }
+        }
+
+        /// <summary>
+        /// Asserts that the input index is valid for a given <#=item.XmlType#> instance.
+        /// </summary>
+        /// <typeparam name="T">The item of items in the input <#=item.XmlType#> instance.</typeparam>
+        /// <param name="index">The input index to be used to access <paramref name="<#=item.Name#>"/>.</param>
+        /// <param name="<#=item.Name#>">The input <#=item.XmlType#> instance to use to validate <paramref name="index"/>.</param>
+        /// <param name="name">The name of the input parameter being tested.</param>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if <paramref name="index"/> is not valid to access <paramref name="<#=item.Name#>"/>.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsInRangeFor<T>(int index, <#=item.Type#> <#=item.Name#>, string name)
+        {
+            if ((uint)index >= (uint)<#=item.Name#>.<#=item.Size#>)
```

Comment:
Need lower-bound check too, eh?

Reply:
Ah, it's a trick! 😄

Actually we don't, because signed integers are represented in [2-complement](https://en.wikipedia.org/wiki/Two%27s_complement) (most significant bit indicates the sign, the others represent the absolute value with inverted bits + 1, with overflow). Checking whether an index is not in a _positive_ `int` range is equivalent to:

```csharp
if (index < 0 || index >= max) throw ...
```

So, if we cast both values to `uint`, we'll get the following:

- If the index was positive, it doesn't change. Then the `>=` comparison is valid, as we want out index to only be _lower_ than the maximum possible index.
- If the index was negative, casting it to an _unsigned_ value makes it so that the most significant bit (the sign) becomes part of the absolute value. Hence, if the value was negative, it'll become `index + int.MaxValue`, which is guaranteed to be always higher than the maximum index, as that number is always positive when read in 2-complement. The same comparison still works, as we want our `uint` cast to be strictly _lower_ than the maximum.

So basically we get two checks for the price of one! 😊
This is the same trick used throughout CoreCLR as well to perform range checks.

The practical effect is that this completely removes the conditional branch in asm 🎉
Here's the version with the two conditions (RyuJit, x64):

```asm
; bool IsInRangeTwo(int i, int max) => i >= 0 && i < max;

C.IsInRangeTwo(Int32, Int32)
    L0000: test ecx, ecx
    L0002: jl L000d
    L0004: cmp ecx, edx
    L0006: setl al
    L0009: movzx eax, al
    L000c: ret
    L000d: xor eax, eax
    L000f: ret

; IsInRangeOne(int i, int max) => (uint)i < (uint)max;

C.IsInRangeOne(Int32, Int32)
    L0000: cmp ecx, edx
    L0002: setb al
    L0005: movzx eax, al
    L0008: ret
```

Shorter, and with no branching! 🚀

Reply:
![image](https://user-images.githubusercontent.com/24302614/76455788-507e0580-6393-11ea-8878-0233c2b248d5.png)

This is what I rely on perf experts like you for 😋. Maybe just add a comment to call out that we're using an optimized trick here for future me (and others)?

Reply:
Absolutely! Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/f17aacd5a1e0f58df35536fb5d15bf9f4dcb0fcd 👍